proposals/0534-metro-package-exports-support.md

```diff
@@ -0,0 +1,455 @@
+
+This default is directly misaligned with react-native-web via Webpack  ([docs](https://webpack.js.org/guides/package-exports/#target-environment)), which is used by a majority of users.
+
+**Proposed**:
+
+- **Breaking**: We will implement the `"browser"` condition name — changed from the top-level field to be selected when the React Native platform is `'web'`.
+- **Breaking**: We will introduce a `"react-native"` condition name — changed from the top-level field to represent a native React Native JS runtime environment (and not web).
+- Metro will not provide an implementation for other community conditions.
+
+Combined, this is the only breaking change we foresee in Metro that will immediately impact app developers, and which we are not deferring via strict mode opt-in. It will align behaviour for existing users of react-native-web via Webpack.
+
+| Condition name | Description | Node.js | Webpack | Metro (proposed) |
+| - | - | - | - | - | - |
+| `"browser"` | Any web browser environment | - | Matched when `platform === 'web'` | Matched when `platform === 'web'` |
+| `"react-native"` [New] | Any React Native runtime environment (non-browser) | - | Matched when `platform !== 'web'` | Matched when `platform !== 'web'` |
+| All other | | (unimplemented) | Various supported, e.g. `"deno"`, `"worker"`, `"electron"` | (unimplemented) |
+| `"default"`| Makes no assumption about JS environment | Always matched | Always matched | Always matched |
+
+Note: Now that `"react-native"` and `"browser"` will not overlap, packages intending to support both native and web environments but which want to expose alternative `"exports"` targeting *React Native **APIs*** should continue to use the `"default"` condition and use a [subpath export](#subpath-exports) instead: e.g. `import { reactNativeExport } from 'some-pkg/react-native'`.
+
+### Conditional exports: User conditions
+
+User conditions relates to the ability to set custom condition names within a project, which may be matched by target packages.
+
+Prior art: [Webpack's `resolve.conditionNames`](https://webpack.js.org/configuration/resolve/#resolveconditionnames) and [Node.js' `--conditions` CLI argument](https://nodejs.org/api/cli.html#cli_c_condition_conditions_condition).
+
+**Proposed**:
+
+- P1 (not targeted in first iteration): We will provide a new config option within `metro.config.js` allowing app developers to define additional condition names which statically extend the default conditions matched.
+
+```js
+module.exports = {
+  // ...
+  resolver: {
+    conditionNames: ['browser', 'production'],
+  },
+};
+```
+
+The exact name and shape of this option may change during implementation.
+
+In addition, `resolver.resolveRequest` will continue to provide an escape hatch from Metro's handling of conditional exports, should an app need to override imports from a given package.
+
+### Opting into strict `"exports"` handling
```

Comment:
if it's _not_ strict by default, #512 is probably even more needed as Jest doesn't allow you to toggle strictness (similar to Node)